app/components/modal.tsx

```diff
@@ -1,75 +1,97 @@
-  function openModal(content: React.ReactNode) {
-    setContent(content);
-    setState("OPEN");
-  }
-
-  function closeModal() {
-    setState("CLOSING");
-  }
-
-  function closeModalFinal() {
-    setContent(undefined);
-    setState("CLOSED");
-  }
+export function Modal(props: {
+  open: boolean;
+  onClose: () => void;
+  children: React.ReactNode;
+  className?: string; // override modal content container style e.g. max width/height
+}) {
+  const { floating, context } = useFloating({
+    open: props.open,
+    onOpenChange: (open) => {
+      tinyassert(!open); // should get only `open = false` via `useDismiss`
+      props.onClose();
+    },
+  });
+  const { getFloatingProps } = useInteractions([useDismiss(context)]);
+  const id = useId();
 
   return (
-    <DefaultModalContext.Provider value={{ openModal, closeModal }}>
-      {children}
-      <Transition appear show={state === "OPEN"}>
-        <div
-          className="absolute inset-0 z-50 flex justify-center items-center"
-          data-test="modal"
-        >
+    <FloatingPortal id={id}>
+      <Transition appear show={props.open} className="z-100">
+        {/* backdrop */}
+        <Transition.Child
+          className="transition duration-300 fixed inset-0 bg-black"
+          enterFrom="opacity-0"
+          enterTo="opacity-40"
+          entered="opacity-40"
+          leaveFrom="opacity-40"
+          leaveTo="opacity-0"
+        />
+        {/* content */}
+        <RemoveScroll className="fixed inset-0 overflow-hidden flex justify-center items-center">
           <Transition.Child
-            className="transition duration-300 absolute inset-0 z-[-1]"
-            enterFrom="opacity-0"
-            enterTo="opacity-100"
-            leaveFrom="opacity-100"
-            leaveTo="opacity-0"
-          >
-            <div
-              className="absolute inset-0 bg-black/[0.3]"
-              onClick={closeModal}
-            />
-          </Transition.Child>
-          <Transition.Child
-            className="w-full max-w-xl" // TODO: options for `openModal`
-            enter="transition duration-300"
+            className={cls(
+              props.className,
+              "transition duration-300 transform w-[90%] max-w-xl shadow-lg"
+            )}
             enterFrom="opacity-0 scale-95"
             enterTo="opacity-100 scale-100"
-            leave="transition duration-300"
             leaveFrom="opacity-100 scale-100"
             leaveTo="opacity-0 scale-95"
-            afterLeave={closeModalFinal}
           >
-            {content}
+            <div
+              {...getFloatingProps({
+                ref: floating,
+                className: "w-full h-full",
+              })}
+            >
+              {props.children}
+            </div>
           </Transition.Child>
-        </div>
+        </RemoveScroll>
       </Transition>
-    </DefaultModalContext.Provider>
+    </FloatingPortal>
   );
-};
+}
+
+export function useModal() {
```

Comment:
Let's backport this idea to https://github.com/hi-ogawa/unocss-preset-antd/blob/main/packages/app/src/components/modal.tsx